spec/lang/well-formed.md

```diff
@@ -114,11 +114,18 @@ impl Type {
             Enum { variants, size, discriminator, discriminant_ty, .. } => {
                 // All the variants need to be well-formed and be the size of the enum so
                 // we don't have to handle different sizes in the memory representation.
-                // Also their alignment may not be larger than the total enum alignment.
+                // Also their alignment may not be larger than the total enum alignment and
+                // all the values written by the tagger must fit into the variant.
+                // FIXME (Timon): should we require alignment checks for the values written in the taggers?
```

Comment:
I don't understand the fixme. Do you mean having a separate notion of alignment for the tagger? That would require changes in the type itself as well? I don't think we want or need that. But we'll have to make sure in minimize that we represent whatever alignment restriction might exist on the discriminant field.

Reply:
Rust does alignment checks upon loading the tag, and forces types like in the example to have 2 bytes alignment.
```rust
enum A {
    A1 = 0,
    A2 = 6000,
}
```

However I think this is mainly because of how some targets need their loads aligned and may be too restrictive to make this a thing in MiniRust.

Reply:
We have already checked that the surrounding place is sufficiently aligned. That also implicitly ensures that the tag is aligned, so we don't need a separate check here.

Reply:
> However I think this is mainly because of how some targets need their loads aligned and may be too restrictive to make this a thing in MiniRust.

MiniRust absolutely needs to check everything required for any target. Otherwise the compiler to the target has to be able to deal with unaligned loads. After all, we can't have code be okay for MiniRust but then fail on the target, that would be a critical bug.

Reply:
No. The offset at which the values are read/written *defines* the alignment requirement.

There are Rust types where a 2-byte discriminant resides at an odd offset, i.e., it is 1-aligned. This is completely fine and the codegen backend knows how to handle it.

Reply:
Oh, okay... The alignment of the tag may only have an influence with direct tagging, I see.

Reply:
No. It's more like, the offset of the tag in the surrounding enum *defines* which alignment guarantees there are, and the compiler computes the LLVM IR alignment annotations from that.